### PR TITLE
feat: integrate storagev2 in building index of segcore

### PR DESCRIPTION
**Pull request by sunby**
issue: https://github.com/milvus-io/milvus/issues/28655

### CODE REVIEW

File: internal/core/src/index/IndexFactory.cpp [hunk 1]
```diff
@@ -54,6 +54,28 @@ IndexFactory::CreateScalarIndex<std::string>(
...
+template <>
+ScalarIndexPtr<std::string>
+IndexFactory::CreateScalarIndex(
+    const IndexType& index_type,
+    const storage::FileManagerContext& file_manager_context,
+    std::shared_ptr<milvus_storage::Space> space) {
+#if defined(__linux__) || defined(__APPLE__)
+    return CreateStringIndexMarisa(file_manager_context, space);
+#else
+    throw std::runtime_error("unsupported platform");
```
Comment (jiaoew1991) [inline internal/core/src/index/IndexFactory.cpp:75]: throw SegcoreError with code

File: internal/core/src/index/IndexFactory.cpp [hunk 2]
```diff
@@ -139,8 +174,82 @@ IndexFactory::CreateVectorIndex(
...
+        case DataType::DOUBLE:
+            return CreateScalarIndex<double>(index_type, file_manager, space);
+
+            // create string index
+        case DataType::STRING:
+        case DataType::VARCHAR:
+            return CreateScalarIndex<std::string>(
+                index_type, file_manager, space);
+        default:
+            throw std::invalid_argument(
```
Comment (jiaoew1991) [inline internal/core/src/index/IndexFactory.cpp:209]: ditto

File: internal/core/src/index/ScalarIndexSort.h [hunk 3]
```diff
@@ -69,7 +80,7 @@ class ScalarIndexSort : public ScalarIndex<T> {
     Range(T lower_bound_value,
           bool lb_inclusive,
           T upper_bound_value,
-          bool ub_inclusive) override;
+          bool ub_inclsive) override;
```
Comment (jiaoew1991) [inline internal/core/src/index/ScalarIndexSort.h:83]: typo?

File: internal/core/src/index/VectorMemIndex.cpp [hunk 4]
```diff
@@ -112,6 +171,101 @@ VectorMemIndex<T>::Load(const BinarySet& binary_set, const Config& config) {
...
+        if (!status.ok()) {
+            PanicInfo(DataFormatBroken, "unable to read slice meta");
+        }
+        auto raw_slice_meta =
+            storage::DeserializeFileData(slice_meta_data, slice_meta_sz);
+        Config meta_data = Config::parse(std::string(
+            static_cast<const char*>(raw_slice_meta->GetFieldData()->Data()),
+            raw_slice_meta->GetFieldData()->Size()));
+        for (auto& item : meta_data[META]) {
+            std::string prefix = item[NAME];
```
Comment (jiaoew1991) [inline internal/core/src/index/VectorMemIndex.cpp:225]: what about use `std::string_view`

File: internal/core/src/index/VectorMemIndex.cpp [hunk 5]
```diff
@@ -112,6 +171,101 @@ VectorMemIndex<T>::Load(const BinarySet& binary_set, const Config& config) {
...
+        }
+    }
+
+    if (!pending_index_files.empty()) {
+        for (auto& file_name : pending_index_files) {
+            auto raw_index_blob = read_blob(file_name);
+            index_datas.insert({file_name, raw_index_blob->GetFieldData()});
+        }
+    }
+    LOG_SEGCORE_INFO_ << "construct binary set...";
```
Comment (jiaoew1991) [inline internal/core/src/index/VectorMemIndex.cpp:252]: info or debug ?

File: internal/core/src/indexbuilder/index_c.cpp [hunk 6]
```diff
@@ -137,6 +145,89 @@ CreateIndex(CIndex* res_index, CBuildIndexInfo c_build_index_info) {
     }
 }
 
+CStatus
+CreateIndexV3(CIndex* res_index, CBuildIndexInfo c_build_index_info) {
```
Comment (jiaoew1991) [inline internal/core/src/indexbuilder/index_c.cpp:149]: CreateIndexV2 already deleted, you can use CreateIndexV2 now

File: internal/core/src/index/IndexFactory.cpp (hunk 1 above)
Comment (sunby) [reply internal/core/src/index/IndexFactory.cpp:75]: I decide to change them in this pr ><

File: internal/core/src/index/VectorMemIndex.cpp (hunk 4 above)
Comment (sunby) [reply internal/core/src/index/VectorMemIndex.cpp:225]: > what about use `�std::string_view`

prefix will be used as string& in a function later, but string_view can not be cast to string implicitly

File: internal/core/src/segcore/SegmentSealedImpl.cpp [hunk 7]
```diff
@@ -221,6 +224,53 @@ SegmentSealedImpl::LoadFieldData(const LoadFieldDataInfo& load_info) {
...
+            load_info.url,
+            milvus_storage::Options{nullptr, load_info.storage_version});
+        AssertInfo(res.ok(),
+                   fmt::format("init space failed: {}, error: {}",
+                               load_info.url,
+                               res.status().ToString()));
+        std::shared_ptr<milvus_storage::Space> space = std::move(res.value());
+        auto load_future = pool.Submit(
+            LoadFieldDatasFromRemote2, space, schema_, field_data_info);
+        LOG_SEGCORE_ERROR_ << "finish submitting LoadFieldDatasFromRemote task "
```
[the remark this answers is not in the document]
Comment (sunby) [reply internal/core/src/segcore/SegmentSealedImpl.cpp:259]: > why is error level?

just for debug, I will change it back

**Pull request merged**